test/images/image-util.sh

```diff
@@ -117,23 +158,75 @@ push() {
+      exit 1
+    fi
+
+    if [[ "$os_name" = "linux" ]]; then
+      docker push "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}"
+    elif [[ -n "${REMOTE_DOCKER_URL:-}" ]]; then
+      # NOTE(claudiub): We're pushing the image we built on the remote Windows node.
+      docker --tlsverify --tlscacert "${HOME}/.docker-${os_version}/ca.pem" \
+        --tlscert "${HOME}/.docker-${os_version}/cert.pem" --tlskey "${HOME}/.docker-${os_version}/key.pem" \
+        -H "${REMOTE_DOCKER_URL}" push "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}-${os_version}"
+    else
+      echo "Cannot push the image '${IMAGE}' for ${os_arch}. REMOTE_DOCKER_URL_${os_version} should be set, containing the URL to a Windows docker daemon."
+    fi
   done
 
+  # NOTE(claudiub): if the REMOTE_DOCKER_URL var is not set, or it is an empty string, we mustn't include
+  # Windows images into the manifest list.
+  if test -z "${REMOTE_DOCKER_URL:-}" && printf "%s\n" "$os_archs" | grep -q '^windows'; then
+    echo "Skipping pushing the image '${IMAGE}' for Windows. REMOTE_DOCKER_URL_\${os_version} should be set, containing the URL to a Windows docker daemon."
+    os_archs=$(printf "%s\n" "$os_archs" | grep -v "^windows")
+  fi
+
   kube::util::ensure-gnu-sed
 
   # The manifest command is still experimental as of Docker 18.09.2
   export DOCKER_CLI_EXPERIMENTAL="enabled"
-  # Make archs list into image manifest. Eg: 'amd64 ppc64le' to '${REGISTRY}/${IMAGE}-amd64:${TAG} ${REGISTRY}/${IMAGE}-ppc64le:${TAG}'
-  while IFS='' read -r line; do manifest+=("$line"); done < <(echo "$archs" | ${SED} -e "s~[^ ]*~$REGISTRY\/$IMAGE\-&:$TAG~g")
+  # Make os_archs list into image manifest. Eg: 'linux/amd64 linux/ppc64le' to '${REGISTRY}/${IMAGE}:${TAG}-linux-amd64 ${REGISTRY}/${IMAGE}:${TAG}-linux-ppc64le'
+  while IFS='' read -r line; do manifest+=("$line"); done < <(echo "$os_archs" | ${SED} "s~\/~-~g" | ${SED} -e "s~[^ ]*~$REGISTRY\/$IMAGE:$TAG\-&~g")
   docker manifest create --amend "${REGISTRY}/${IMAGE}:${TAG}" "${manifest[@]}"
-  for arch in ${archs}; do
-    docker manifest annotate --arch "${arch}" "${REGISTRY}/${IMAGE}:${TAG}" "${REGISTRY}/${IMAGE}-${arch}:${TAG}"
+  for os_arch in ${os_archs}; do
+    if [[ $os_arch =~ .*/.*/.* ]]; then
+      # for Windows, we have to support both LTS and SAC channels, so we're building multiple Windows images.
+      # the format for this case is: OS/ARCH/OS_VERSION.
+      os_name=$(echo "$os_arch" | cut -d "/" -f 1)
+      arch=$(echo "$os_arch" | cut -d "/" -f 2)
+      os_version=$(echo "$os_arch" | cut -d "/" -f 3)
+      suffix="$os_name-$arch-$os_version"
+    elif [[ $os_arch =~ .*/.* ]]; then
+      os_name=$(echo "$os_arch" | cut -d "/" -f 1)
+      arch=$(echo "$os_arch" | cut -d "/" -f 2)
+      suffix="$os_name-$arch"
+    else
+      echo "The BASEIMAGE file for the ${IMAGE} image is not properly formatted. Expected entries to start with 'os/arch', found '${os_arch}' instead."
+      exit 1
```

Comment:
almost identical code in build() / push(), better to extract into a function?

Reply:
I don't think it will help much. Most of this code you're mentioning is just extracting $os_name, $arch, and $os_version, and setting $REMOTE_DOCKER_URL for windows image, which are still needed either way.